ExternalBeamPlanning/Widgets/qSlicerDoseEngineLogic.cxx

```
@@ -479,3 +489,51 @@ vtkMRMLRTBeamNode* qSlicerDoseEngineLogic::createBeamInPlan(vtkMRMLRTPlanNode* p
 
   return beamNode;
 }
+
+//---------------------------------------------------------------------------
+vtkMRMLRTBeamNode* qSlicerDoseEngineLogic::createArcBeamInPlan( vtkMRMLRTPlanNode* planNode, 
+  double initialAngle, double finalAngle, double stepAngle, bool rotationDirection)
+{
+  // Set beams logic
+  vtkSlicerBeamsModuleLogic* beamsLogic = nullptr;
+  qSlicerAbstractCoreModule* beamsModule = qSlicerCoreApplication::application()->moduleManager()->module("Beams");
+  if (beamsModule)
+  {
+    beamsLogic = vtkSlicerBeamsModuleLogic::SafeDownCast(beamsModule->logic());
+  }
+  else
+  {
+    qCritical() << Q_FUNC_INFO << ": Beams module is not found";
+  }
+
+  if (!planNode || !planNode->GetScene())
+  {
+    qCritical() << Q_FUNC_INFO << ": Invalid plan node!";
+    return nullptr;
+  }
+
+  vtkNew<vtkMRMLSequenceNode> beamSequenceNode;
+  vtkNew<vtkMRMLSequenceNode> transformSequenceNode;
+  vtkNew<vtkMRMLSequenceBrowserNode> beamSequenceBrowserNode;
+
+  if (beamsLogic->CreateArcBeamDynamicSequence( initialAngle, finalAngle, stepAngle,
+    rotationDirection, planNode, beamSequenceBrowserNode, beamSequenceNode, transformSequenceNode))
+  {
+    // Get proxy beam node
+    vtkMRMLNode* node = beamSequenceBrowserNode->GetProxyNode(beamSequenceNode);
+    vtkMRMLRTBeamNode* proxyBeamNode = vtkMRMLRTBeamNode::SafeDownCast(node);
+
+    // Get proxy transform node
+    node = beamSequenceBrowserNode->GetProxyNode(transformSequenceNode);
+    vtkMRMLLinearTransformNode* proxyTransformNode = vtkMRMLLinearTransformNode::SafeDownCast(node);
+
+    if (proxyBeamNode && proxyTransformNode)
+    {
+      // Add proxy beam node to plan (create Subject Hierarchy for beam node)
+      planNode->AddProxyBeam(proxyBeamNode);
+      proxyBeamNode->SetAndObserveTransformNodeID(proxyTransformNode->GetID());
+      return proxyBeamNode;
+    }
+  }

```

Should report an error if creation failed?